### DISCUSSION TITLE
How to handle upgrade of dependencies from breaking pinned postgres?

### DISCUSSION THREAD

**Discussion by richardkmichael** [Homebrew/discussions#1438] [Everyday usage]
I have `postgresql@12` pinned.  Postgresql has several dependencies, including `icu4c`.  When I upgrade packages, `postgresql` is correctly kept at version 12, however homebrew updates `icu4c`.  This breaks postgres because it can no longer find the required, dynamically loaded, `icu4c` version.

This has happened several times.  It is fairly frustrating, especially because it seems counter to the point of pinning `postgresql`.

I suspect the breakage is dependent on how "far" icu4c jumps ahead.  Patch version bumps probably still satisfy the linker.  In this case, icu4u went from 0.68.x to 0.69.x, and so dyld is unhappy.

I assume I must be doing something wrong.  I have read the documentation; searched GitHub issues and discussions; and StackOverflow and haven't found a clear solution.

What is the "homebrew way" of solving this problem?  

Note, I do not care about increased disk space usage.  I am fine with brew keeping an "extra" version of all the postgresql dependencies, a copy of each at whatever version is required by the pinned postgresql.

What I've considered:

- I found a suggestion to [use `HOMEBREW_NO_INSTALL_CLEANUP`](https://github.com/Homebrew/homebrew-core/issues/39567#issuecomment-490229955).  I would be inclined to `pin` all the postgresql dependencies, but that same issue  comment recommends against it.  I am sure this would cause many other packages to also not update, but I require postgresql to remain working, so I might live with it.

- I could also probably force a re-install of the same version of postgresql.  IIRC, I've tried this in the past, and something else broke.  (Admittedly vague, sorry.)

Last, could this be a bug in the postgresql formula?

Relevant versions:

macOS 10.14.6

`brew --version`
Homebrew 3.1.5-101-gd3013fc
Homebrew/homebrew-core (git revision 18512fab18; last commit 2021-05-07)

`brew ls --versions postgresql@12`
postgresql@12 12.6_1

Thanks for any help.

**Comment by SMillerDev** [comment]
> What is the "homebrew way" of solving this problem?

The homebrew way would be accepting that homebrew is an evergreen package manager and using the latest version. Short of that, using https://formulae.brew.sh/formula/postgresql@12 instead of https://formulae.brew.sh/formula/postgresql seems like a good solution.

**Reply by richardkmichael** [reply]
I was already using `postgresql@12`, as I mentioned in my post.  (I edited it just now to try and make that clearer.)

As I need postgres to remain stable, I've uninstalled Homebrew postgresql and switched to Postgres.app instead.  This way I'll have better control over updating.

Thank you!

**Reply by carlocab** [reply]
Do you also need a specific minor and patch version for `postgresql@12`? If you don't, there is no need to pin it. Updates to `postgresql@12` whenever `icu4c` is updated are almost exclusively just the exact same version of PostgreSQL 12, just rebuilt against the new `icu4c`, so there should be no danger in those upgrades (i.e. they are the exact same upstream version).
